pkg/asset/machines/azure/machines.go

```diff
@@ -0,0 +1,107 @@
+	azureprovider "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1alpha1"
+)
+
+const (
+	// TODO(flaper87): We're choosing to hardcode these values to make
+	// the environment more predictable. We expect there to a secret
+	// named `azure-credentials` and a cloud named `AzurePublicCloud` in
+	// the clouds file stored in this secret.
+	cloudsSecret          = "azure-credentials"
+	cloudsSecretNamespace = "kube-system"
+
+	// CloudName is a constant containing the name of the cloud used in the internal cloudsSecret
+	// TODO: parameterized this to support non-public cloud regions and azure stack
+	CloudName = "AzurePublicCloud"
+)
+
+// Machines returns a list of machines for a machinepool.
+func Machines(clusterID string, config *types.InstallConfig, pool *types.MachinePool, osImage, role, userDataSecret string) ([]machineapi.Machine, error) {
+	if configPlatform := config.Platform.Name(); configPlatform != azure.Name {
+		return nil, fmt.Errorf("non-Azure configuration: %q", configPlatform)
+	}
+	if poolPlatform := pool.Platform.Name(); poolPlatform != azure.Name {
+		return nil, fmt.Errorf("non-Azure machine-pool: %q", poolPlatform)
+	}
+	platform := config.Platform.Azure
+	mpool := pool.Platform.Azure
+
+	total := int64(1)
+	if pool.Replicas != nil {
+		total = *pool.Replicas
+	}
+	var machines []machineapi.Machine
+	for idx := int64(0); idx < total; idx++ {
+		az := ""
+		provider, err := provider(clusterID, platform, mpool, osImage, az, role, userDataSecret)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create provider")
+		}
+		machine := machineapi.Machine{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "machine.openshift.io/v1beta1",
+				Kind:       "Machine",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "openshift-machine-api",
+				Name:      fmt.Sprintf("%s-%s-%d", clusterID, pool.Name, idx),
+				Labels: map[string]string{
+					"machine.openshift.io/cluster-api-cluster":      clusterID,
+					"machine.openshift.io/cluster-api-machine-role": role,
+					"machine.openshift.io/cluster-api-machine-type": role,
+				},
+			},
+			Spec: machineapi.MachineSpec{
+				ProviderSpec: machineapi.ProviderSpec{
+					Value: &runtime.RawExtension{Object: provider},
+				},
+				// we don't need to set Versions, because we control those via operators.
+			},
+		}
+
+		machines = append(machines, machine)
+	}
+
+	return machines, nil
+}
+
+func provider(clusterID string, platform *azure.Platform, mpool *azure.MachinePool, osImage string, az string, role, userDataSecret string) (*azureprovider.AzureMachineProviderSpec, error) {
+	return &azureprovider.AzureMachineProviderSpec{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "azureprovider.k8s.io/v1alpha1",
+			Kind:       "AzureMachineProviderSpec",
+		},
+		UserDataSecret:    &corev1.SecretReference{Name: userDataSecret},
+		CredentialsSecret: &corev1.SecretReference{Name: "azure-credentials", Namespace: "kube-system"},
```

Comment:
this is supposed to be in `openshift-machine-api` right? like https://github.com/openshift/installer/blob/0137c5460221c26ee7d966014c907372926188d1/pkg/asset/machines/aws/machines.go#L97

Reply:
can you explain? this is the provider spec, same for aws that you link to.

Reply:
machines use the local secret object in namespace `openshift-machine-api` `aws-cloud-credentials`  on AWS that is generated with smaller and audited permissions. So I was asking the question if machine-api-operator will be setting up such secret for azure too in that namespace.

hopefully that helps.

Reply:
> this is supposed to be in openshift-machine-api right?
it wasn't clear you were talking about the namespace. I actually don't know. @awesomenix has been working on the machine-api part and implemented this.
again, we have been working on getting the cluster up and running, not on all these details.

Reply:
@abhinavdahiya is machine api doing operations similar to cred minter? I thought https://github.com/openshift/cloud-credential-operator handled all credential management. Right not for azure its all pass through and will not be using functionality from machine api.

Reply:
I see, https://github.com/openshift/machine-api-operator/blob/2dfd35e0c7af141037942fdefadd6eb84c567587/install/0000_30_machine-api-operator_00_credentials-request.yaml#L10 

makes sense for the request to delegate this to credential manager instead of doing pass through here.

We do not have credential manager implemented yet, hence this workaround :(